README.md
# decision_tree_lab

New csv(New_Eval.csv) and json(New_Eval.json) for Course_Eval(1) in public_html



history.txt
python -m http.server
php
python -m http.server 8080
cd ..
dir
cd public_html
cd ..
python lab_course.py
cd public_html
python C:\Users\akars\Desktop\decision_tree_lab-master\public_html\data\NewEval.csv a a
python lab_course.py
ls
dir
cd ..
dir
python test_student_eval.py C:\Users\akars\Desktop\decision_tree_lab-master\public_html\data\NewEval.csv asdf fuckme
cd public_html
python -m http.server 8080
git
cd ..
ls
dir
cwd
wd
cd ..
cd decision_tree_lab-master
git status
cd..
git clone https://github.com/gitkarsh/decision_tree_lab.git
dir
cd decision_tree_lab
dir
git add all
git add -A
git commit -m "done"
git push
cd public_html
dir
python -m http.server 8080
cd ..
git add -A
git commit -m "done with One"
git push
git pull
git push
git pull
git add -A
git commit -m "done with One"
git push
doskey /HISTORY > history.txt


test_student_eval.py
import dtree_build
import sys
import csv


def main(col_names=None):
    # parse command-line arguments to read the name of the input csv file
    # and optional 'draw tree' parameter
    if len(sys.argv) < 2:  # input file name should be specified
        print ("Please specify input csv file name")
        return

    csv_file_name = sys.argv[1]

    data = []
    with open(csv_file_name) as csvfile:
        readCSV = csv.reader(csvfile, delimiter=',')
        for row in readCSV:
            data.append(list(row))

    print("Total number of records = ",len(data))
    tree = dtree_build.buildtree(data, min_gain =0.01, min_samples = 5)

    dtree_build.printtree(tree, '', col_names)

    max_tree_depth = dtree_build.max_depth(tree)
    print("max number of questions=" + str(max_tree_depth))

    if len(sys.argv) > 2: # draw option specified
        pass
        # import dtree_draw
        # dtree_draw.drawtree(tree, jpeg=csv_file_name+'.jpg')

    if len(sys.argv) > 3:  # create json file for d3.js visualization
        import json
        import dtree_to_json
        json_tree = dtree_to_json.dtree_to_jsontree(tree, col_names)
        print(json_tree)

        # create json data for d3.js interactive visualization
        with open(csv_file_name + ".json", "w") as write_file:
            json.dump(json_tree, write_file)


if __name__ == "__main__":
    col_names = ['rank',
                 'ethnicity',
                 'gender',
                 'language',
                 'age',
                 'class_size',
                 'cls_level',
                 'bty_avg',
                 'prof_eval']
    main(col_names)







Course_Eval/lab_course.py
import math


def same(x):  # For attributes which we don't change
    return x


def course_evaluation(x):  # Course evaluation attribute
    if x >= 4:
        return "good"

    return "bad"


def age(x):  # Age attribute
    if x <= 30:
        return 30
    if x <= 40:
        return 40
    if x <= 50:
        return 50
    if x <= 60:
        return 60
    if x <= 70:
        return 70

    return (math.ceil(x/10))*10


def class_size(x):  # Class size attribute
    if x <= 30:
        return 30
    if x <= 50:
        return 50
    if x <= 100:
        return 100
    if x <= 200:
        return 200

    return 300


def beauty_score(x):  # Beauty score attribute
    if x <= 3:
        return 3
    if x > 6:
        return 7

    return math.ceil(x)


def prof_evaluation(x):  # Professor evaluation score attribute
    if x >= 4:
        return 4
    if x >= 3:
        return 3
    if x >= 2:
        return 2


# List of functions to execute corresponding to each attribute
change_fns = [same, same, same, same, age, class_size, same, beauty_score, prof_evaluation, course_evaluation]


def main(readname, writename):
    readfile = open(readname)
    writefile = open(writename, "w")

    c = 0
    for line in readfile:
        if c != 0:
            writefile.write("\n")

        # Get array of attribute values
        attributes = line.rstrip().split(",")

        # Iterate through attribute values
        for i in range(len(attributes)):
            attribute = attributes[i]

            try:
                attribute = float(attribute)
            except ValueError:
                pass

            writefile.write(str(change_fns[i](attribute)) + ("," if i != (len(attributes) - 1) else ""))

        c = 1


if __name__ == "__main__":
    main("StudentEvaluations.csv - StudentEvaluations.csv.csv", "../NewEval.csv")

Regression/regr_dtree_draw.py
from PIL import Image, ImageDraw
from regression_tree import decisionnode

def getwidth(tree):
    if tree.tb == None and tree.fb == None: return 1
    return getwidth(tree.tb) + getwidth(tree.fb)


def getdepth(tree):
    if tree.tb == None and tree.fb == None: return 0
    return max(getdepth(tree.tb), getdepth(tree.fb)) + 1


def drawtree(tree, jpeg='tree.jpg'):
    w = getwidth(tree) * 100
    h = getdepth(tree) * 100 + 120

    img = Image.new('RGB', (w, h), (255, 255, 255))
    draw = ImageDraw.Draw(img)

    drawnode(draw, tree, w / 2, 20)
    img.save(jpeg, 'JPEG')


def drawnode(draw, tree, x, y):
    if tree.results == None:
        # Get the width of each branch
        w1 = getwidth(tree.fb) * 100
        w2 = getwidth(tree.tb) * 100

        # Determine the total space required by this node
        left = x - (w1 + w2) / 2
        right = x + (w1 + w2) / 2

        # Draw the condition string
        draw.text((x - 20, y - 10), str(tree.col)[0:5] + ':' + str(tree.value), (0, 0, 0))

        # Draw links to the branches
        draw.line((x, y, left + w1 / 2, y + 100), fill=(255, 0, 0))
        draw.line((x, y, right - w2 / 2, y + 100), fill=(255, 0, 0))

        # Draw the branch nodes
        drawnode(draw, tree.fb, left + w1 / 2, y + 100)
        drawnode(draw, tree.tb, right - w2 / 2, y + 100)
    else:
        txt = ' \n' + str(tree.results)
        draw.text((x - 20, y), txt, (0, 0, 0))

Regression/regr_lab_course.py
import math


def same(x):  # For attributes which we don't change
    return x


def course_evaluation(x):  # Course evaluation attribute
    if x >= 4:
        return "good"

    return "bad"


def age(x):  # Age attribute
    if x <= 30:
        return 30
    if x <= 40:
        return 40
    if x <= 50:
        return 50
    if x <= 60:
        return 60
    if x <= 70:
        return 70

    return (math.ceil(x/10))*10


def class_size(x):  # Class size attribute
    if x <= 30:
        return 30
    if x <= 50:
        return 50
    if x <= 100:
        return 100
    if x <= 200:
        return 200

    return 300


def beauty_score(x):  # Beauty score attribute
    if x <= 3:
        return 3
    if x > 6:
        return 7

    return math.ceil(x)


def prof_evaluation(x):  # Professor evaluation score attribute
    if x >= 4:
        return 4
    if x >= 3:
        return 3
    if x >= 2:
        return 2


# List of functions to execute corresponding to each attribute
change_fns = [same, same, same, same, age, class_size, same, beauty_score, prof_evaluation, course_evaluation]


def main(readname, writename):
    readfile = open(readname)
    writefile = open(writename, "w")

    c = 0
    for line in readfile:
        if c != 0:
            writefile.write("\n")

        # Get array of attribute values
        attributes = line.rstrip().split(",")

        # Iterate through attribute values
        for i in range(len(attributes)):
            attribute = attributes[i]

            try:
                attribute = float(attribute)
            except ValueError:
                pass

            writefile.write(str(change_fns[i](attribute)) + ("," if i != (len(attributes) - 1) else ""))

        c = 1


if __name__ == "__main__":
    main("StudentEvaluations.csv - StudentEvaluations.csv.csv", "../NewRegr.csv")

Regression/regression_tree.py
from math import log

class decisionnode:
    def __init__(self, col=-1, value=None, results=None, tb=None, fb=None):
        self.col = col
        self.value = value
        self.results = results
        self.tb = tb  # True branch
        self.fb = fb  # False branch


# Divides a set on a specific column. Can handle numeric
# or nominal values
def divideset(rows, column, value):
    # Make a function that tells
    # if the split is on numeric value (>=/<)
    # or on nominal value (==/!=)
    split_function = None
    if isinstance(value, int) or isinstance(value, float):
        split_function = lambda row: row[column] >= value
    else:
        split_function = lambda row: row[column] == value

    # Divide the rows into two sets and return them
    set1 = [row for row in rows if split_function(row)]
    set2 = [row for row in rows if not split_function(row)]
    return (set1, set2)


# Create counts of class labels for a given set
# (the last column of each row is the class attribute)
def uniquecounts(rows):
    results = {}
    for row in rows:
        # The result is the last column
        r = row[len(row) - 1]
        if r not in results:
            results[r] = 0
        results[r] += 1
    return results


# Probability that a randomly placed item will
# be in the wrong category
# Using the closed form here:
# https://en.wikipedia.org/wiki/Decision_tree_learning#Gini_impurity
def giniimpurity(rows):
    total = len(rows)
    counts = uniquecounts(rows)
    imp = 0
    for k1 in counts:
        p1 = float(counts[k1]) / total
        for k2 in counts:
            if k1 == k2:
                continue
            p2 = float(counts[k2]) / total
            imp += p1 * p2
    return imp


# Entropy is the sum of p(x)log(p(x)) across all
# the different possible classes
def entropy(rows):
    results = uniquecounts(rows)
    ent = 0.0
    for r in results.keys():
        p = float(results[r]) / len(rows)
        ent = ent - p * log(p, 2)
    return ent


def variance(rows):
    if len(rows) == 0: return 0
    data = [float(row[len(row) - 1]) for row in rows]
    mean = sum(data) / len(data)
    variance = sum([(d - mean) ** 2 for d in data]) / len(data)
    return variance





def prediction(leaf_labels):
    total = 0
    result = {}
    for label, count in leaf_labels.items():
        total += count
        result[label] = count

    for label, val in result.items():
        result[label] = str(int(result[label]/total * 100))+"%"

    return result


def classify(observation, tree):
    if tree.results != None:
        return str(tree.results)
    else:
        v = observation[tree.col]
        branch = None
        if isinstance(v, int) or isinstance(v, float):
            if v >= tree.value:
                branch = tree.tb
            else:
                branch = tree.fb
        else:
            if v == tree.value:
                branch = tree.tb
            else:
                branch = tree.fb
        return classify(observation, branch)


# Classify an observation with missing data
def mdclassify(observation, tree):
    if tree.results != None:
        return prediction(tree.results)
    else:
        v = observation[tree.col]
        if v == None:
            tr, fr = mdclassify(observation, tree.tb), mdclassify(observation, tree.fb)
            tcount = sum(tr.values())
            fcount = sum(fr.values())
            tw = float(tcount) / (tcount + fcount)
            fw = float(fcount) / (tcount + fcount)
            result = {}
            for k, v in tr.items(): result[k] = v * tw
            for k, v in fr.items(): result[k] = v * fw
            return result
        else:
            if isinstance(v, int) or isinstance(v, float):
                if v >= tree.value:
                    branch = tree.tb
                else:
                    branch = tree.fb
            else:
                if v == tree.value:
                    branch = tree.tb
                else:
                    branch = tree.fb
            return mdclassify(observation, branch)


def buildtree(rows, scoref=variance,
              min_gain=0, min_samples=0):
    if len(rows) == 0:
        return decisionnode()
    current_score = scoref(rows)

    # Set up accumulator variables to track the best criteria
    best_gain = 0.0
    best_criteria = None
    best_sets = None

    column_count = len(rows[0]) - 1
    for col in range(0, column_count):
        # Generate the list of different values in
        # this column
        column_values = {}
        for row in rows:
            column_values[row[col]] = 1
        # Now try dividing the rows up for each value
        # in this column
        for value in column_values.keys():
            (set1, set2) = divideset(rows, col, value)

            # Information gain
            p = float(len(set1)) / len(rows)
            gain = current_score - p * scoref(set1) - (1 - p) * scoref(set2)
            if gain > best_gain and len(set1) > min_samples and len(set2) > min_samples and gain > min_gain:
                best_gain = gain
                best_criteria = (col, value)
                best_sets = (set1, set2)

    # Create the sub branches
    if best_gain > 0:
        trueBranch = buildtree(best_sets[0], scoref, min_gain, min_samples)
        falseBranch = buildtree(best_sets[1], scoref, min_gain, min_samples)
        return decisionnode(col=best_criteria[0], value=best_criteria[1],
                            tb=trueBranch, fb=falseBranch)
    else:
        leafMean = sum(map(lambda x: float(x[len(x) - 1]), rows))/len(rows)
        return decisionnode(results=round(leafMean*10)/10)


def max_depth(tree):
    if tree.results != None:
        return 0
    else:
        # Compute the depth of each subtree
        tDepth = max_depth(tree.tb)
        fDepth = max_depth(tree.fb)

        # Use the larger one
        if (tDepth > fDepth):
            return tDepth + 1
        else:
            return fDepth + 1


def printtree(tree, current_branch, attributes=None,  indent=''):
    # Is this a leaf node?
    if tree.results != None:
        print(indent + current_branch + str(tree.results))
    else:
        # Print the split question
        split_col = str(tree.col)
        if attributes is not None:
            split_col = attributes[tree.col]
        split_val = str(tree.value)
        if type(tree.value) == int or type(tree.value) == float:
            split_val = ">=" + str(tree.value)
        print(indent + current_branch + split_col + ': ' + split_val + '? ')

        # Print the branches
        indent = indent + '  '
        printtree(tree.tb, 'T->', attributes, indent)
        printtree(tree.fb, 'F->', attributes, indent)

Regression/regr_dtree_to_json.py
import json
import regression_tree
import sys
import csv


def node_label(node, attributes=None):
    if node.results != None:
        return str(node.results)
    split_col = str(node.col)
    if attributes is not None:
        split_col = attributes[node.col]
    split_val = str(node.value)
    if type(node.value) == int or type(node.value) == float:
        split_val = ">=" + str(node.value)
    return split_col + ': ' + split_val + '?'


def get_child(node, attributes=None):
    if node == None:
        return None
    curr_node = {}
    curr_node["name"] = node_label(node, attributes)
    children = []
    tb_child = get_child(node.tb, attributes)
    if tb_child:
        children.append(tb_child)
    fb_child = get_child(node.fb, attributes)
    if fb_child:
        children.append(fb_child)

    if len(children) > 0:
        curr_node["children"] = children
    return curr_node


def dtree_to_jsontree(tree, attributes=None):
    root = {}
    root["name"] = node_label(tree, attributes)
    children = []
    tb_child = get_child(tree.tb, attributes)
    if tb_child:
        children.append(tb_child)
    fb_child = get_child(tree.fb, attributes)
    if fb_child:
        children.append(fb_child)

    root["children"] = children
    return root


def main(col_names=None):
    # parse command-line arguments to read the name of the input csv file
    if len(sys.argv) < 2:  # input file name should be specified
        print("Please specify input csv file name")
        return

    csv_file_name = sys.argv[1]

    data = []
    with open(csv_file_name) as csvfile:
        readCSV = csv.reader(csvfile, delimiter=',')
        for row in readCSV:
            data.append(list(row))
    tree = regression_tree.buildtree(data, min_gain =0.01, min_samples = 5)

    max_tree_depth = regression_tree.max_depth(tree)
    print("max number of questions=" + str(max_tree_depth))

    json_tree = dtree_to_jsontree(tree, col_names)
    print(json_tree)

    # create json data for d3.js interactive visualization
    with open(csv_file_name + ".json", "w") as write_file:
        json.dump(json_tree, write_file)


if __name__ == "__main__":
    col_names = [
        'Age',
        'Work class',
        'Education',
        'Marital status',
        'Occupation',
        'Relationship',
        'Race',
        'Sex',
        'Native-country',
        'Income']
    main(col_names)

Regression/regr_dtree_example_small.py
import regr_dtree
import sys

if __name__ == "__main__":
    # fruits with their size and color
    fruits = [
        [4, 'red', 'apple'],
        [4, 'green', 'apple'],
        [1, 'red', 'cherry'],
        [1, 'green', 'grape'],
        [5, 'red', 'apple']
    ]

    tree = regr_dtree.buildtree(fruits)
    regr_dtree.printtree(tree, '', ["size", "color"])
    print("fruit [2, 'red'] is: ", regr_dtree.classify([2, 'red'], tree))
    print("fruit [4.5, 'red'] is: ", regr_dtree.classify([4.5, 'red'], tree))
    print("fruit [1.4, 'green'] is: ", regr_dtree.classify([1.4, 'green'], tree))

    max_tree_depth = regr_dtree.max_depth(tree)
    print("max number of questions="+str(max_tree_depth))
    if len(sys.argv) > 1: # draw option specified
        import dtree_draw
        dtree_draw.drawtree(tree, jpeg='fruits_dt.jpg')

Regression/regr_test_student_eval.py
import regression_tree
import sys
import csv

def self_test(tree):
    data = [["tenure track", "minority", "female", "english", "50", "30", "upper", "7", "4"],
            ["teaching", "not minority", "male", "english", "30", "30", "lower", "5", "3"],
            ["tenure track", "minority", "male", "english", "40", "30", "lower", "5", "4"]]

    for row in data:
        print(row, end="")
        print(" prediction = " + regression_tree.classify(row, tree))

def main(col_names=None):
    # parse command-line arguments to read the name of the input csv file
    # and optional 'draw tree' parameter
    if len(sys.argv) < 2:  # input file name should be specified
        print ("Please specify input csv file name")
        return

    csv_file_name = sys.argv[1]

    data = []
    with open(csv_file_name) as csvfile:
        readCSV = csv.reader(csvfile, delimiter=',')
        for row in readCSV:
            data.append(row)

    print("Total number of records = ",len(data))
    tree = regression_tree.buildtree(data, min_gain =0.003, min_samples = 5)

    regression_tree.printtree(tree, '', col_names)

    max_tree_depth = regression_tree.max_depth(tree)
    print("max number of questions=" + str(max_tree_depth))

    if len(sys.argv) > 2: # draw option specified
        import regression_tree_draw
        regression_tree_draw.drawtree(tree, jpeg=csv_file_name+'.jpg')

    if len(sys.argv) > 3:  # create json file for d3.js visualization
        import json
        import regression_tree_to_json
        json_tree = regression_tree_to_json.dtree_to_jsontree(tree, col_names)
        print(json_tree)

        # create json data for d3.js interactive visualization
        with open(csv_file_name + ".json", "w") as write_file:
            json.dump(json_tree, write_file)

    self_test(tree)


if __name__ == "__main__":
    col_names = ['rank',
                 'ethnicity',
                 'gender',
                 'language',
                 'age',
                 'class_size',
                 'cls_level',
                 'bty_avg',
                 'prof_eval']
    main(col_names)